spmd/__init__.py

```diff
@@ -1,106 +1,96 @@
-    if placements is None:
-        placements = [Replicate() for _ in range(device_mesh.ndim)]
 
-    # distribute the tensor according to DTensorSpec
-    for idx, placement in enumerate(placements):
-        if placement.is_shard():
-            placement = cast(Shard, placement)
-            shard_dim = placement.dim
-            assert (
-                shard_dim <= tensor.ndim
-            ), f"Sharding dim {shard_dim} greater than tensor ndim {tensor.ndim}"
-            # TODO: handle uneven shard sizes
-            num_chunks = device_mesh.size(dim=idx)
-            assert tensor.size(shard_dim) % num_chunks == 0, (
-                f"Only support chunk sharding evenly now, but tensor got "
-                f"dimension {shard_dim} of size {tensor.size(shard_dim)}, "
-                f"which does not divide number of shards {num_chunks}."
-            )
-            chunk_size = tensor.size(shard_dim) // num_chunks
-            tensor_list = list(tensor.chunk(num_chunks, dim=shard_dim))
-            scatter_shape = list(tensor.size())
-            scatter_shape[shard_dim] = chunk_size
-            local_tensor = device_mesh.scatter(tensor_list, mesh_dim=idx)
-            # scatter call could not return a tensor with correct requires_grad
-            # field, as ProcessGroupNCCL refuse to take a tensor with requires_grad
-            # to do inplace update! So we manually set it here
-            local_tensor.requires_grad_(tensor.requires_grad)
-            tensor = local_tensor
-        elif placement.is_replicate():
-            tensor = device_mesh.broadcast(tensor, mesh_dim=idx)
-        else:
-            raise RuntimeError("Not supported!")
+    if device_mesh is None:
+        device_mesh = get_global_device_mesh()
 
-    return DTensor(
-        tensor,
-        device_mesh,
-        placements,
-        requires_grad=tensor.requires_grad,
+    # set this to true on demand, to avoid inplace update the parameter directly
+    # This is purely for nn.Module API parameter replacement BC reason
+    overwrite_on_conversion = (
+        torch.__future__.get_overwrite_module_params_on_conversion()
     )
+    torch.__future__.set_overwrite_module_params_on_conversion(True)
 
-
-# pyre-fixme[3]: Return type must be annotated.
-def distribute_module(
-    mod: nn.Module,
-    device_mesh: Optional[DeviceMesh] = None,
-    spec: Optional[Sequence[Placement]] = None,
-):
-    """
-    this function coverts all module parameters
-    to distributed tensor parameters according to
-    the placements and device_mesh spcified.
-    TODO: add a more flexible tagging, i.e. convert
-    certain param to a certain spec, like a PlacementPlan
-    """
-
-    # pyre-fixme[3]: Return type must be annotated.
-    # pyre-fixme[2]: Parameter must be annotated.
-    def to_dist_tensor(t):
-        if isinstance(t, nn.Parameter):
-            return distribute_tensor(t.data, device_mesh, spec)
+    # this function loop over the whole module parameters
```

Comment:
Can this be a function docstring?

Reply:
This is a function we used internally in `distribute_module`, I guess it does not need a docstring unless we want to make this a preset function that's visible to the user?